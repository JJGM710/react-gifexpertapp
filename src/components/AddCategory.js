import React, { useState } from 'react'
import PropTypes from 'prop-types';

const AddCategory = ({setCategories, categories}) => {

    const [inputValue,setInputValue]= useState('')

    const handleInputChange = (e) => {
        //console.log(e.target.value);
        setInputValue(e.target.value)


        console.log('llamdo de evento onchange')
    }

    const handleSubmit = (e) =>{
        e.preventDefault();
        //console.log('handleSubmit', inputValue)
        let test = categories.find((ctg) => ctg.toLowerCase() === inputValue.trim().toLowerCase());
        if (inputValue.trim().length > 2 && test === undefined)
        {
            setCategories(ctg => [inputValue, ...ctg]);
            console.log('se hizo el cambio')
            setInputValue('');
        }
       // console.log(categories);
        
    }
    return (
        <form onSubmit={handleSubmit}>
            <p> {inputValue} </p>
            <input 
                placeholder="Search.."
                className='box'
                type = "text" 
                value = {inputValue}
                onChange = {handleInputChange}
            ></input>
        </form>
    )
}

AddCategory.propTypes = {
    setCategories : PropTypes.func.isRequired,
    categories: PropTypes.array.isRequired
};

export default AddCategory;
