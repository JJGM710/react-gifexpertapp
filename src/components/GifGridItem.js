import React from 'react'
import { PropTypes } from 'prop-types';

export const GifGridItem = ({id, title,url}) => {
   // console.log(id, title, url);
    return (
        <div className="card animate__animated animate__fadeIn animate__delay-1s">
            <img src={url} alt={title} />
            <p> {title?title.toLowerCase():'titulo no definido'} </p>
        </div>
    )
}


//definicon de default props y de isrequired


//validacion de propiedades
GifGridItem.propTypes = {
    title : PropTypes.string.isRequired,
    url : PropTypes.string.isRequired
};

//definir propiedades por defect
GifGridItem.defaultProps = {
};


// 1.enzyme
// 2.enzyme to jSON
// 3. debe mostar el comp correctamente
    
// shallow
// wrapper 
// expect wrapper toMatchSnapshot 
