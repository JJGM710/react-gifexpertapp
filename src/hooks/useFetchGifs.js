import { useState, useEffect } from 'react';
import { getGifs } from './../helpers/getGifs';
// recibe los datos de imgs, para guardarlos en data y tambien se indica cuando termina de cargar los archivos
export const useFetchGifs = (category) => {
    const [state, setstate] = useState({
        data:[],
        loading: true,
    });

    //obterner cuando la categoria cambia, use effect  no puede ser async
    useEffect(() => {

            getGifs(category)
            .then(imgs =>{

                setstate({
                    data: imgs,
                    loading:false
                })
            })



    },[category])

    return state; //retorna el usesate como estado inciial
}