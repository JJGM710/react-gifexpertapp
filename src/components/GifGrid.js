import React from 'react'
import { GifGridItem } from './GifGridItem';
//import { getGifs } from './../helpers/getGifs';
import { useFetchGifs } from './../hooks/useFetchGifs';

export const GifGrid = ({category}) => {

    // const [images, setimages] = useState([]);

    // // genera una dependecia con category, es decir, que cada vez que el valor de esa propiedad cambie va a renderizar ese component
    // useEffect(() => {
    //     getGifs(category)
    //         .then(imgs => setimages(imgs));
    // }, [category])

    const {data : images, loading} = useFetchGifs(category);


    //getGifs();
    return (
        <>
        <h3 className="animate__animated animate__flash animate__slow">{category}</h3>

        {loading && <h2 className="animate__animated animate__flash animate__slower">Loading...</h2>}

        <div className= 'card-grid'>
            
            {images.map((img) => (
                 <GifGridItem 
                    key = {img.id}
                    {...img} />
            ))
            }

        </div>
        </>
    );
};
