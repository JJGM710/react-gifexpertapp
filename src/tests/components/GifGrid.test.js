import  React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import { GifGrid } from './../../components/GifGrid';
import { useFetchGifs } from './../../hooks/useFetchGifs';
jest.mock('./../../hooks/useFetchGifs');

describe('Pruebas en GifGrid', () => {

    const category = 'One Punch Man'

    test('debe renderizar el componente en su estado inicial', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        const wrapper = shallow(<GifGrid category={category}/>);
        expect(wrapper).toMatchSnapshot();
        
    })
    
    test('debe de mostrar items cuando se cargan imagenes con useFetchGifs', () => {
        
        const imgs = [{
            id: '12dsad3',
            title: 'Probando Gifs',
            url: 'https://safs.asdsad.com'
        },
        {
            id: '12dsa213d3',
            title: 'Probando Gifs2',
            url: 'https://safs.asdsdsad.com'
        }
        ]
        
        useFetchGifs.mockReturnValue({
            data: imgs,
            loading: false
        });

        const {id, title, url}=imgs[0];
        console.log(id)
        const wrapper = shallow(<GifGrid category={category}/>);

        expect(wrapper.find('p').exists()).toBe(false)
        expect(wrapper.find('GifGridItem').length).toBe(imgs.length)
       
    })
    
})
