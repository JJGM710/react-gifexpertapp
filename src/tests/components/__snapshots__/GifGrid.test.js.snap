// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifGrid debe renderizar el componente en su estado inicial 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__flash animate__slow"
  >
    One Punch Man
  </h3>
  <h2
    className="animate__animated animate__flash animate__slower"
  >
    Loading...
  </h2>
  <div
    className="card-grid"
  />
</Fragment>
`;
